cmd/geoipupdate/main.go

```diff
@@ -0,0 +1,428 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	flock "github.com/theckman/go-flock"
+)
+
+// Version is the program's version number.
+const Version = "0.0.1"
+
+func main() {
+	log.SetFlags(0)
+
+	args := getArgs()
+
+	config, err := NewConfig(args.ConfigFile, args.DatabaseDirectory)
+	if err != nil {
+		log.Fatalf("Error loading configuration file: %+v", err)
+	}
+
+	lock, err := setup(config, args.Verbose)
+	if err != nil {
+		log.Fatalf("Error preparing to update: %+v", err)
+	}
+	defer func() {
+		if err := lock.Unlock(); err != nil {
+			log.Fatalf("Error unlocking lock file: %+v", errors.Wrap(err, "unlocking"))
+		}
+	}()
+
+	if err := run(config, args.Verbose); err != nil {
+		log.Fatalf("Error retrieving updates: %+v", err)
+	}
+}
+
+func setup(
+	config *Config,
+	verbose bool,
+) (*flock.Flock, error) {
+	if err := maybeSetProxy(config); err != nil {
+		return nil, err
+	}
+
+	if err := checkEnvironment(config); err != nil {
+		return nil, err
+	}
+
+	lock := flock.NewFlock(config.LockFile)
+	ok, err := lock.TryLock()
+	if err != nil {
+		return nil, errors.Wrap(err, "error acquiring a lock")
+	}
+	if !ok {
+		return nil, errors.Errorf("could not acquire lock on %s", config.LockFile)
+	}
+	if verbose {
+		log.Printf("Acquired lock file lock (%s)", config.LockFile)
+	}
+
+	return lock, nil
+}
+
+var client = &http.Client{
+	Timeout: 30 * time.Minute, // Allow for slow downloads
+}
+
+func maybeSetProxy(config *Config) error {
+	if config.Proxy == "" {
+		return nil
+	}
+	proxy := config.Proxy
+
+	if strings.Index(proxy, ":") == -1 {
+		proxy += ":1080" // The 1080 default came from cURL.
+	}
+	proxy = "http://" + proxy
+
+	proxyURL, err := url.Parse(proxy)
+	if err != nil {
+		return errors.Wrap(err, "error parsing proxy URL")
+	}
+
+	if config.ProxyUserPassword != "" {
+		userPassword := strings.SplitN(config.ProxyUserPassword, ":", 2)
+		if len(userPassword) != 2 {
+			return errors.New("proxy user/password is malformed")
+		}
+		proxyURL.User = url.UserPassword(userPassword[0], userPassword[1])
+	}
+
+	http.DefaultTransport.(*http.Transport).Proxy = http.ProxyURL(proxyURL)
+
+	return nil
+}
+
+func checkEnvironment(
+	config *Config,
+) error {
+	fi, err := os.Stat(config.DatabaseDirectory)
+	if err != nil {
+		return errors.Wrap(err, "database directory is not available")
+	}
+
+	if !fi.IsDir() {
+		return errors.New("database directory is not a directory")
+	}
+
+	// I don't think there is a reliable cross platform way to check the
+	// directory is writable. We'll discover that when we try to write to it
+	// anyway.
+
+	return nil
+}
+
+func run(
+	config *Config,
+	verbose bool,
+) error {
+	for _, editionID := range config.EditionIDs {
```

Comment:
A future improvement might be that we could do these in parallel, although the gain on that might not be that great as it would mostly be network-bound.

Reply:
Yeah, that would be cool!